File: Design-Patterns-Using-Kotlin/Mediator/MediatorPattern2.kt
interface Mediator {
    fun addColleague(colleague: Colleague)
    fun sendMessage(message: String, colleagueId: String)
}

class ConcreteMediator : Mediator {
    private val colleagues = mutableMapOf<String, Colleague>()

    override fun addColleague(colleague: Colleague) {
        colleagues[colleague.id] = colleague
    }

    override fun sendMessage(message: String, colleagueId: String) {
        println("Mediator pass Message: $message")
        colleagues[colleagueId]?.receive(message)
    }
}

interface Colleague {
    val id: String
    fun send(message: String, to: String)
    fun receive(message: String)
}

class ConcreteColleague(override val id: String, private val mediator: Mediator) : Colleague {
    override fun send(message: String, to: String) {
        println("$id Sent Message: $message")
        mediator.sendMessage(message, to)
    }

    override fun receive(message: String) {
        println("$id Received Message: $message")
    }
}

// Client Code
fun main() {
    val mediator: Mediator = ConcreteMediator()
    val first = ConcreteColleague("First", mediator)
    mediator.addColleague(first)

    val second = ConcreteColleague("Second", mediator)
    mediator.addColleague(second)

    first.send("Hello, World!", "Second")
    second.send("Hi, World!", "First")
}

/*
First Sent Message: Hello, World!
Mediator pass Message: Hello, World!
Second Received Message: Hello, World!
Second Sent Message: Hi, World!
Mediator pass Message: Hi, World!
First Received Message: Hi, World!
*/

File: Design-Patterns-Using-Ruby/Mediator/MediatorPattern2.rb
# Mediator Interface
class Mediator
    def add_colleague(colleague)
        raise NotImplementedError, 'Subclasses must implement add_colleague'
    end

    def send_message(message, colleague_id)
        raise NotImplementedError, 'Subclasses must implement send_message'
    end
end

# Concrete Mediator
class ConcreteMediator < Mediator
    def initialize
        @colleagues = {}
    end

    def add_colleague(colleague)
        @colleagues[colleague.id] = colleague
    end

    def send_message(message, colleague_id)
        puts "Mediator pass Message: #{message}"
        @colleagues[colleague_id].receive(message)
    end
end

# Colleague Abstract Class
class Colleague
    attr_reader :id, :mediator

    def initialize(id, mediator)
        @id = id
        @mediator = mediator
    end

    def send(message, to)
        raise NotImplementedError, 'Subclasses must implement send'
    end

    def receive(message)
        raise NotImplementedError, 'Subclasses must implement receive'
    end
end

# Concrete Colleague
class ConcreteColleague < Colleague
    def send(message, to)
        puts "#{id} Sent Message: #{message}"
        mediator.send_message(message, to)
    end

    def receive(message)
        puts "#{id} Received Message #{message}"
    end
end

# Client Code
mediator = ConcreteMediator.new
first = ConcreteColleague.new('First', mediator)
mediator.add_colleague(first)

second = ConcreteColleague.new('Second', mediator)
mediator.add_colleague(second)

first.send('Hello, World!', 'Second')
second.send('Hi, World!', 'First')

=begin 
First Sent Message: Hello, World!
Mediator pass Message: Hello, World!
Second Received Message Hello, World!
Second Sent Message: Hi, World!
Mediator pass Message: Hi, World!
First Received Message Hi, World!
=end